src/outputs/parthenon_hdf5.cpp

```diff
@@ -274,45 +279,132 @@ void PHDF5Output::WriteOutputFile(Mesh *pm, ParameterInput *pin, SimTime *tm) {
+      for (int i = 0; i < BOUNDARY_NFACES; i++) {
+        boundaryConditions.push_back(GetBoundaryString(pm->mesh_bcs[i]));
+      }
+      writeH5ASTRINGS(
+          "BoundaryConditions", boundaryConditions,
+          H5S::FromHIDCheck(H5Screate_simple(1, boundaryConditions_size, NULL)),
+          infoDSet);
+
+      // DatasetNames - which datasets from the top level to read
+      const hsize_t datasetNames_size[1] = {output_params.variables.size()};
+      writeH5ASTRINGS("DatasetNames", output_params.variables,
+                      H5S::FromHIDCheck(H5Screate_simple(1, datasetNames_size, NULL)),
+                      infoDSet);
+
+      // NumVariables - number of variables within each dataset
+      std::vector<int> numVariables(output_params.variables.size());
+      // VariablesNames - Names of variables within each dataset
+      std::vector<std::string> variableNames;
+
+      for (int i = 0; i < ciX.vars.size(); i++) { // for every block
+        const std::vector<std::string> component_labels =
+            ciX.vars[i]->metadata().getComponentLabels();
+
+        if (component_labels.size() > 0) {
+          numVariables[i] = component_labels.size();
+          for (const auto &label : component_labels) {
+            variableNames.push_back(label);
+          }
+        } else {
+          numVariables[i] = 1;
+          variableNames.push_back(ciX.vars[i]->label());
+        }
+      }
+
+      writeH5AI32("NumVariables", numVariables.data(),
+                  H5S::FromHIDCheck(H5Screate_simple(1, datasetNames_size, NULL)),
+                  infoDSet);
+
+      const hsize_t variableNames_size[1] = {variableNames.size()};
+      writeH5ASTRINGS("VariableNames", variableNames,
+                      H5S::FromHIDCheck(H5Screate_simple(1, variableNames_size, NULL)),
+                      infoDSet);
+
+      for (const auto &package : pm->packages.AllPackages()) {
+        std::shared_ptr<StateDescriptor> state = package.second;
+        // Write IncludeWithOutputParams
+        std::vector<std::string> includeWithOutputParams =
+            state->IncludeWithOutputParams();
+        for (const auto &param_name : includeWithOutputParams) {
+          const std::type_index type = state->ParamType(param_name);
+
+          std::stringstream ss;
+          ss << state->label() << "/" << param_name;
+
+          if (type == typeid(int)) {
+            const int &param = state->Param<int>(param_name);
+            writeH5AI32(ss.str().c_str(), &param, localDSpace, infoDSet);
+          } else if (type == typeid(double)) {
+            const double &param = state->Param<double>(param_name);
+            writeH5AF64(ss.str().c_str(), &param, localDSpace, infoDSet);
+          } else if (type == typeid(std::string)) {
+            const std::string &param = state->Param<std::string>(param_name);
+            writeH5ASTRING(ss.str().c_str(), param, localDSpace, infoDSet);
+          }
```

Comment:
We should probably have `uint32_t`, `uint64_t`, `int64_t`, and `float` here as well. This will become a bit easier with my PR #486 that makes writing attributes a template function on the type.

Reply:
Yes. I only added the types that already had a convenient macro to write the attribute. In your PR with the template function, we can add the other obvious types.

Reply:
I'm fine with this.